core/src/main/kotlin/org/evomaster/core/search/algorithms/MosaAlgorithm.kt

```diff
@@ -32,51 +42,73 @@ class MosaAlgorithm<T> : SearchAlgorithm<T>() where T : Individual {
 
             //new generation
 
-            val nextPop: MutableList<EvaluatedIndividual<T>> = mutableListOf()
+            val nextPop: MutableList<Data> = mutableListOf()
+
+            while (nextPop.size < n-1) {
 
-            while (nextPop.size < n) {
+                var ind = selection()
 
-                getMutatator().mutateAndSave(selection(), archive)
-                        ?.let { nextPop.add(it) }
+                var eval = getMutatator().mutateAndSave(ind, archive)
+
+                nextPop.add(Data(eval as EvaluatedIndividual))
 
                 if (!time.shouldContinueSearch()) {
                     break
                 }
             }
+            // generate one random solution
+            var ie = sampleIndividual()
+            nextPop.add(Data(ie as EvaluatedIndividual))
 
             population.addAll(nextPop)
             sortPopulation()
-            while (population.size > n) {
-                population.removeAt(population.size - 1)
-            }
         }
 
         return archive.extractSolution()
     }
 
-    private class Data(val ind: EvaluatedIndividual<*>) {
-
-        var rank = -1
-        var crowdingDistance = -1
-        var dominationByCounter = 0
-        var dominatedSolutions: MutableList<Data> = mutableListOf()
-    }
-
 
     private fun sortPopulation() {
 
         val notCovered = archive.notCoveredTargets()
-        val list = population.map { ind -> Data(ind) }
 
-        mosaPreferenceCriterion(notCovered, list)
+        val fronts = preferenceSorting(notCovered, population)
 
-        fastNonDominatedSort(notCovered, list)
+        var remain: Int = config.populationSize
+        var index = 0
+        population.clear()
 
-        subvectorDominance(notCovered, list)
+        // Obtain the next front
+        var front = fronts[index]
+
+        while (front!=null && remain > 0 && remain >= front.size && front.isNotEmpty()) {
+            // Assign crowding distance to individuals
+            subvectorDominance(notCovered, front)
+            // Add the individuals of this front
+            for (d in front)
```

Comment:
please, all blocks should be within "{}", eg "for(d in front){...}"
I should add something like a "Checkstyle" to enforce those rules...

Reply:
Right!